README.md
# darkwebmysterybot

**Dark Web Mystery Bot v5.4 stable (20220604)**

Dark Web Mystery Bot, created by redneonglow, is a Fediverse phrase bot inspired by YouTube videos and urban legends about the Deep Web. The bot generates silly phrases such as these:

* dark net mystery programming
* marianas web mystery republican
* hoax mystery box
* troll.clos
* marianas net salvadorian
* dark web mystery chemist
* 3 AM moroccan potion
* deep web taekwondo box
* overnight algebra challenge 
* dark net netball
* deep web sausage
* 24 hour basketball challenge

Phrases are derived from the following lists:

* christmas (christmas, winter holidays)^%
* conspiracy (conspiracy theories, pseudoscience, aliens, the occult)^
* italian (italian food, italian-american foods, pizza toppings)
* medieval (stuff from fairy tales)^
* nationalities (nationalities, ethnicities, races)^
* occupations (originally based on the studies list below)
* politics (related to us politics)^
* sports (professional sports, extreme sports, martial arts)
* studies (academic studies)

^ excluded from safe-space mode

% only active during month of December

Dark Web Mystery Bot can post directly to Mastodon and Pleroma instances and is great for use in an hourly cronjob.

WARNING: Dark Web Mystery Bot may produce phrases which are only suitable for "free speech" instances. For "safe space" instances, an option is provided to exclude such possibilities, but there are still no guarantees.

**REQUIREMENTS:**

* Python 3.6 or higher
* Mastodon.py and its dependencies
* curl

**SET UP THE TOKEN FILE:**

1. Create a Fediverse account for Dark Web Mystery Bot.
2. Set up a token here: https://tinysubversions.com/notes/mastodon-bot/
3. Create a token file by running this command:
   `curl <command you are given> > tokenfile.json`

Note that if you change the password on the account, you will need to create a new token file.

**VISIBILITY:**

The visibility option may be any of the following:

* `direct` (only visible to the bot account)
* `private` (only visible to the bot account's followers)
* `public` (visible to everyone)
* `unlisted` (visible to everyone, but hidden from the public timeline)

In most cases you would want to use either `public` or `unlisted` for the visibility option.

**EXAMPLE COMMANDS:**

Show help: `./dwmbot.py -h`

Show license (Simplified BSD): `./dwmbot.py -l`

Show version: `./dwmbot.py -v`

Print 1 phrase to stdout: `./dwmbot.py -p 1`

Print 4 phrases to stdout: `./dwmbot.py -p 4`

Post to an account on Pleroma instance bae.st using token file tokenfile.json and public visibility:

`./dwmbot.py -o https://bae.st tokenfile.json public`

Post to an account on Mastodon instance mastodon.social, using token file wokenfile.json and unlisted visibility, while avoiding phrases which would not be suitable for safe-space instances:

`./dwmbot.py -c https://mastodon.social wokenfile.json unlisted`

Post version info to an account on Pleroma instance bae.st using token file tokenfile.json (visibility always unlisted):

`./dwmbot.py -w https://bae.st tokenfile.json`

Enjoy!

-redblade7 aka redneonglow

**SPECIAL THANKS**

* Your New SJW Waifu (`@sjw@bae.st`) for allowing me to host the bot on bae.st
* Dielan (`@dielan@shitposter.club`) for showing me how to create Fediverse bots with Python
* chincostud (`@asylore@bae.st`) who suggested adding shadow web/shadow net support

**FEDIVERSE CONTACT INFO:**

* `@redneonglow@bae.st` / https://bae.st/redneonglow (main)
* `@redneonglow@weeaboo.space` / https://weeaboo.space/redneonglow (backup)
* `@redneonglow@gameliberty.club` / https://gameliberty.club/@redneonglow (backup)
* `@redneonglow@poa.st` / https://poa.st/@redneonglow (backup)

The author runs an instance of Dark Web Mystery Bot here, generating phrases every half hour:

* `@darkweb@bae.st` / https://bae.st/darkweb


CHANGES
v5.4 stable (20220604)

* Update copyright, contact info, and version
* Remove already added feature in TODO

v5.3 stable (20211220)

* Update changelog file

v5.2 stable (20211218)

* Replace neckbeard.xyz with bae.st
* Update contact information

v5.1 stable (20210328)
* Removed duplicate "elf" from christmas list (already in medieval)
* Updated copyright dates to 2021 (forgot to do this in v5)

v5 stable (20210112)
Summary of changes since v4:
* Add some QAnon-related words to conspiracy
* Updated documentation to be more specific about required Python version
* Bot account on Neckbeard now generates every half hour
* Removed "etc." from conspiracy list in documentation
* Add christmas list (only activated during December)

v5-dev (20201028)
* Removed "etc." from conspiracy list in documentation
* Added christmas list (only activated during December)

v5-dev (20200706)
* Updated documentation to be more specific about required Python version
* Bot account on Neckbeard now generates every half hour

v5-dev (20200516)
* Add some QAnon-related words to conspiracy

v4 stable (20200403)
Summary of changes since v3:
* Added lots more words
* Added note in README.md about changing passwords and token files
* Corrected typo in comment
* Added support for boxes (mystery box, dybbuk box, etc)
* Added support for challenges (24 hour challenge, 3 AM challenge, etc)
* Remove duplicate "surveillance" from studies (was already in conspiracy)
* Remove unneeded plurals

v4-dev (20200322)
* Added challenges
* Added more to conspiracy and politics

v4-dev (20200321)
* Added more to conspiracy, italian, medieval, and politics

v4-dev (20200217)
* Added note in README.md about changing passwords and token files

v4-dev (20200124)
* Correct typo in comment

v4-dev (20200121)
* Added a less edgy synonym to nationalities (old one is still there)
* Added more to conspiracy and nationalities
* Added a lot more to politics
* Remove unneeded plurals

v4-dev (20200120)
* Added a synonym to nationalities
* Added a few more to conspiracy, italian, nationalities, occupations, politics, and studies
* Removed duplicate "surveillance" from studies (was already in conspiracy)

v4-dev (20200119)
* Added support for boxes

v3 stable (20191208)
Summary of changes since v2.1:
* Add visibility support (introduces a new required command line parameter - remember to update your crontabs with this version!)
* Added potion support
* Added generation of fake dark web domains (incl .i2p, .onion, and urban legend TLDs)
* Added "shadow web"/"shadow net" support (idea by `@asylore@neckbeard.xyz`)
* Added "sports" list
* Added more to conspiracy
* Added CONTRIBUTING.md with contact information for suggestions
* Added CODE_OF_CONDUCT.md with contents "Dark Web Mystery Bot exists for the sole purpose of entertainment."

v3-dev (20191205)
* Added "shadow web"/"shadow net" support (idea by `@asylore@neckbeard.xyz`)

v3-dev (20191203)
* Added more to conspiracy

v3-dev (20191202-r1)
* Fixed bug with safe space mode

v3-dev (20191202)
* Comment added to dwmbot.py
* Put lists in alphabetical order in the file
* Add visibility support (introduces a new required command line parameter - remember to update your crontabs with this version!)

v3-dev (20191127)
* Spun off separate pickword() function in preparation for new possibilities
* Added potion support
* Added generation of fake dark web domains (incl .i2p, .onion, and urban legend TLDs)
* Added "sports" list
* Redid probabilities

v2.1 stable (20191126)
Summary of changes since v1:
* Added "safe space mode" which avoids posting politically incorrect content
* Added option to post version in unlisted mode
* Added "nationalities" list (nationalities, ethnicities, races)
* Added "occupations" list (based on the studies list)
* Added more content to the lists

v2-dev (20191119)
* Added "nationalities" list (nationalities, ethnicities, races)
* Added more to conspiracy and medieval

v2-dev (20191117)
* Typo fixed in README.md

v2-dev (20191116)
* Added "occupations" list (based on the studies list)
* Added some more phrases to conspiracy, medieval, politics, and studies 

v2-dev (20191115)
* Created -c/--postpcphrase option (aka "safe space mode")
* Made some more changes to functions to make this possible

v2-dev (20191114)
* Added "Thanks" section to README.md
* Added CHANGES file

dwmbot.py
#!/usr/bin/env python3
#
# Dark Web Mystery Bot
# 
# A Fediverse phrase bot inspired by YouTube videos and urban legends about
# the Deep Web.
#
# Copyright (c) 2019-22, redneonglow
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this
#    list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

from datetime import datetime
from mastodon import Mastodon,MastodonError
import argparse,json,secrets,sys

#program version
progver = "5.4 stable (20220604)"

#prefix
prefix = ["dark net","dark web","deep web","marianas net","marianas web","shadow net","shadow web"]

#faketld (onion, i2p, and fake TLDs associated with urban legends)
faketld = ["clos","dafy","end","i2p","loky","nept","onion","taur"]

#christmas (christmas, winter holidays)
christmas = ["advent","angel","bell","bethlehem","candle","candycane","cake","card","carol","chestnut","chimney","christmas","christmaslist","christmastree","cider","coal","cookie","december","dressing","eggnog","eid","festival","festivus","frankincense","frost","frosty","fruitcake","garland","gift","gold","greeting","grinch","hanukkah","holiday","holly","hollyjolly","ice","icy","jinglebell","jolly","kwanzaa","light","list","manger","menorah","merry","miracle","mistletoe","myrrh","nativity","noel","northpole","ornament","pageant","party","pie","pine","poinsettia","present","pudding","reindeer","ribbon","rudolph","sack","santa","scrooge","season","sheep","shepherd","sled","sleigh","sleighbell","sleighride","snow","snowman","solstice","star","stocking","tannenbaum","tinsel","toy","turkey","tree","vacation","winter","wrapping","wreath","xmas","yule","yulelog","yuletide"]

#conspiracy (conspiracy theories, pseudoscience, aliens, the occult)
conspiracy = ["agenda","adrenochrome","alchemy","alien","antichrist","apocalypse","armageddon","assassination","astrology","aura","autism","awakening","bioweapon","birther","bodycount","brainwashing","cabal","camp","chemtrails","clairvoyance","classified","conspiracy","control","creationism","creepypasta","cryptozoology","deathpanel","deepstate","delusion","demolition","demon","denier","denial","detention","ectoplasm","epidemic","extermination","extraterrestrial","fakenews","fallacy","falseflag","fluoride","foreknowledge","freemasonry","gaybomb","gayfrog","genocide","ghost","ghosthunting","glyphosate","guidestones","hallucination","homeopathy","horoscope","illuminati","infowar","infowarrior","interrogation","intelligence","intuition","jesuit","magick","masonry","massacre","medication","mercury","missing","mkultra","mindcontrol","mindreading","monarch","nibiru","nostradamus","occultism","orb","order","paranoia","paranormal","pizzagate","plot","poison","population","possession","power","precognition","projection","prophecy","prophet","pseudohistory","pseudolaw","pseudoscience","psychic","psycho","psychosis","psychotic","qanon","radiation","reincarnation","remote","reptilian","ritual","revisionism","roswell","satan","satanism","sandyhook","seance","secret","shaman","silverstein","soros","soy","soyboy","sovereign","specter","spirit","spiritualism","supernatural","superstition","surveillance","tapwater","telekinesis","telepathy","thimerosal","tinfoil","torture","truther","undead","vaccination","vaccine","vampire","viewing","zionist","zombie"]

#italian (italian food, italian-american foods, pizza toppings)
italian = ["alfredo","anchovy","artichoke","arugula","asiago","bacon","basil","bologna","bread","broccoli","calamari","cannoli","capicola","cheese","chicken","egg","eggplant","fettucine","garlic","gelato","gorgonzola","jalapeno","ham","lasagna","linguine","macaroni","meatball","meatlovers","mozzarella","mushroom","neapolitan","olive","onion","oregano","parmesan","pasta","peperoncini","pepper","pepperoni","pineapple","pizza","prosciutto","provolone","ravioli","ricotta","risotto","romano","salami","sausage","seafood","shellfish","sicilian","spaghetti","spinach","supreme","tomato","tortellini","ziti"]

#medieval (stuff from fairy tales)
medieval = ["amulet","apprentice","banishment","banshee","beanstalk","bear","beast","bridge","carpet","castle","cauldron","chalice","changeling","crossbow","crown","crusader","crusades","crystal","curse","damsel","dragon","drawbridge","dungeonmaster","dwarf","elf","fairy","fantasy","folktale","frog","genie","giant","gnome","goblin","godmother","gown","grail","guillotine","hag","hangman","harem","hero","heroine","imp","inquisitor","jester","joust","howl","kettle","king","kingdom","knight","lady","land","legend","leprechaun","lord","lore","magic","magician","mace","maiden","mask","mirror","monk","monster","moon","myth","necromancer","newt","nun","ogre","orc","paladin","pegasus","pixie","plague","prince","princess","protector","quail","queen","quest","rack","realm","reign","robe","roundtable","sage","sandman","scepter","scroll","shapeshifter","slippers","soothsayer","sorcerer","spear","spell","spider","sprite","staff","stepmother","sultan","sword","swordfight","talisman","toad","torch","tower","troll","unicorn","valiant","valor","vanquish","vizier","wail","wand","warlock","warrior","werewolf","witch","wizard","wolf"]

#nationalities (nationalities, ethnicities, races)
nationalities = ["afghan","african","albanian","algerian","american","andorran","angolan","argentine","argentinian","armenian","asian","australian","austrian","azerbaijani","bahamian","bangladeshi","barbadian","belarusian","belgian","beninese","bhutanese","black","bolivian","bosnian","brazilian","brit","briton","bruneian","bulgarian","burmese","burundian","cambodian","cameroonian","canadian","caucasian","chadian","chilean","chinese","colombian","congolese","croat","croatian","cuban","cypriot","czech","dane","dominican","dutchman","dutchwoman","ecuadorian","egyptian","emirati","englishman","englishwoman","eritrean","estonian","ethiopian","european","fijian","filipino","finn","frenchman","frenchwoman","gabonese","gambian","georgian","german","ghanaian","greek","grenadian","guatemalan","guinean","guyanese","haitian","hawaiian","hispanic","honduran","hungarian","icelander","indian","indonesian","iranian","iraqi","irishman","irishwoman","islander","israeli","italian","ivorian","jamaican","japanese","jew","jewish","jordanian","kazakh","kazakhistani","kenyan","korean","kuwaiti","laotian","latvian","lebanese","liberian","libyan","liechtensteiner","lithuanian","luxembourger","macedonian","madagascan","malagasy","malawian","malaysian","maldivian","malian","maltese","mauritanian","mauritian","mexican","moldovan","monacan","mongolian","montenegrin","moroccan","mozambican","namibian","native","nepalese","nicaraguan","nigerian","nigerien","norwegian","omani","pakistani","palestinian","panamanian","paraguayan","persian","peruvian","pole","portuguese","qatari","romanian","russian","rwandan","salvadoran","salvadorean","salvadorian","saudi","scot","senegalese","serb","serbian","singaporean","slovak","slovenian","somali","somalian","spaniard","sudanese","surinamese","swazi","swede","swiss","syrian","taiwanese","tadzhik","tajik","tanzanian","thai","togolese","trinidadian","tunisian","turk","turkmen","ugandan","ukrainian","uruguayan","uzbek","venezuelan","vietnamese","welshman","welshwoman","white","yemeni","yugoslav","zambian","zimbabwean"]

#occupations (originally based on the studies list)
occupations = ["accountant","administrator","agent","analyzer","anatomist","anesthesiologist","anthropologist","archeologist","artist","astronaut","astronomer","astrophysicist","biologist","biochemist","botanist","cardiologist","chemist","climatologist","cosmologist","criminologist","cybernetician","dancer","dentist","dermatologist","designer","developer","director","doctor","ecologist","economist","endocrinologist","engineer","ethicist","evolutionist","gastroenterologist","geneticist","geographer","gynecologist","hematologist","hepatologist","historian","linguist","logician","manager","marketer","mathematician","metaphysicist","mechanist","musician","nanotechnologist","nephrologist","neurologist","neurosurgeon","nurse","nutritionist","oncologist","operator","ophthalmologist","optometrist","paleontologist","pathologist","pediatrician","pharmacist","pharmacologist","philosopher","physician","podiatrist","priest","producer","programmer","psychiatrist","psychologist","psychoanalyst","pulmonologist","rheumatologist","robot","scientist","sociologist","spy","statistician","surgeon","sysadmin","technician","theologian","theorist","therapist","toxicologist","writer","zoologist"]

#politics (related to us politics)
politics = ["abortion","abuse","agency","agreement","amendment","alt-left","alt-right","antifa","antitrust","arbitration","argument","arms","assault","backdoor","barack","based","bathroom","bernie","bias","biden","bigotry","bill","bisexuality","bluestate","breakup","bush","capital","capitalism","capitol","cartel","case","catholic","catholicism","censorship","centrist","cheney","china","choice","christian","christianity","church","climate","clinton","college","collusion","communism","confederate","confederacy","congress","congressman","congresswoman","consent","conservative","conservatism","coronavirus","correctness","corruption","constitution","court","coverage","covid","crime","criminal","cuck","debate","democracy","democrat","deplatforming","discrimination","druglord","drugs","drugwar","election","elector","execution","executive","extremist","fascism","fearmonger","feminism","fentanyl","fracking","freedom","gab","gay","gender","gore","guncontrol","guns","hamas","harassment","hate","hatemonger","hatred","heteronormative","heterosexual","hillary","hoax","holocaust","homophobia","homosexuality","house","hydrofracture","illegal","illegals","immigrant","immigration","impeachment","incel","income","inclusion","indecency","insurance","intersex","isis","islam","israel","judiciary","justice","law","left","leftist","legislature","lesbian","liberal","liberalism","libertarian","libertarianism","liberty","maga","marijuana","marriage","masculinity","media","medicare","mexico","microaggression","military","minority","moderate","muslim","narcotics","nazi","neocon","neoconservative","neoconservatism","news","obama","obscenity","opinion","opioid","palestine","partisan","party","patriarchy","peace","pedophilia","penalty","pence","pepe","piracy","police","policy","politician","politics","pornography","presidency","president","privacy","progressive","progressivism","property","polyamory","queer","race","racism","racist","rape","rapist","redpill","redpilled","redstate","religion","reparations","representative","republican","resources","right","rocketman","salary","sanders","scandal","senate","senator","sex","sexuality","sexist","sexism","sjw","slave","slavery","snowflake","socialist","socialism","socialsecurity","speech","state","strawman","tariff","tax","taxation","terror","terrorism","terrorist","theft","trade","transgender","toxic","transphobia","treaty","trump","truth","undocumented","unlawful","vote","voter","voting","wage","wall","war","warming","washington","woke","wuhan","xenophobia"]

#sports (professional sports, extreme sports, martial arts)
sports = ["aikido","archery","athletics","badminton","baseball","basketball","biking","billiards","bouldering","bowling","boxing","canoeing","climbing","cricket","curling","cycling","darts","diving","equestrian","fencing","fishing","football","gliding","golf","gymnastics","handball","hockey","hunting","jai-alai","judo","jiu-jitsu","jumping","karate","kickboxing","motocross","motorsport","mountaineering","netball","ninjutsu","olympics","paragliding","parasailing","pentathlon","pool","racing","riding","rollerblading","rollerskating","rowing","running","rugby","sailing","sandboarding","shooting","skateboarding","skating","skiing","skydiving","snooker","snowboarding","soccer","softball","squash","sumo","surfing","swimming","taekwondo","tennis","volleyball","weightlifting","wrestling","yoga"]

#studies (academic studies)
studies = ["accounting","administration","aerobics","aerodynamics","aerospace","algebra","analysis","anatomy","anesthesiology","anthropology","archeology","art","arts","astronomy","astrophysics","biology","biochemistry","botany","business","calculus","cardiology","chemistry","climatology","computing","cosmology","criminology","cybernetics","dance","dentistry","dermatology","design","development","ecology","economics","electronics","endocrinology","energy","engineering","ethics","evolution","film","finance","fitness","forensics","gastroenterology","genetics","geography","geometry","geriatrics","gynecology","hematology","hepatology","history","humanities","language","languages","linguistics","literature","logic","management","marketing","mathematics","medicine","metaphysics","macroeconomics","mechanics","microeconomics","music","nanotechnology","nephrology","neurology","neurosurgery","nursing","nutrition","obstetrics","oncology","operations","ophthalmology","optometry","paleontology","pathology","pediatrics","pharmacology","pharmaceuticals","philosophy","physics","podiatry","probability","programming","psychiatry","psychology","psychoanalysis","pulmonology","radiology","rehabilitation","rheumatology","robotics","science","security","sociology","statistics","surgery","systems","technology","television","theater","theology","theory","therapy","thermodynamics","toxicology","trigonometry","writing","zoology"]

#returns a random word from a random list
#off: allow offensive content (bool)
def pickword(off):

    if datetime.today().month == 12:
        december = 1
    else:
        december = 0

    if off:
        listname = secrets.randbelow(8 + december)
    else:
        listname = secrets.randbelow(4 + december)

    if listname == 0:
        listword = secrets.choice(italian)
    elif listname == 1:
        listword = secrets.choice(studies)
    elif listname == 2:
        listword = secrets.choice(occupations)
    elif listname == 3:
        listword = secrets.choice(sports)
    elif listname == 4 and off:
        listword = secrets.choice(conspiracy)
    elif listname == 5 and off:
        listword = secrets.choice(politics)
    elif listname == 6 and off:
        listword = secrets.choice(medieval)
    elif listname == 7 and off:
        listword = secrets.choice(nationalities)
    elif listname == 8 and off and december == 1:
        listword = secrets.choice(christmas)
    else:
        print("ERROR: Invalid set!")
        sys.exit(1)

    return str(listword)

#returns the phrase as a string
#off: allow offensive content (bool)
def genphrase(off):
    
    outtype = secrets.randbelow(7)

    if outtype == 0:
        phrase = secrets.choice(prefix) + ' ' + pickword(off)
    elif outtype == 1:
        phrase = secrets.choice(prefix) + " dybbuk " + pickword(off)
    elif outtype == 2:
        phrase = secrets.choice(prefix) + " mystery " + pickword(off)
    elif outtype == 3:
        phrase = pickword(off) + '.' + secrets.choice(faketld)
    elif outtype == 4:
        #keep chance of 3 AM the same as the other potions
        potiontype = secrets.randbelow(len(prefix) + 1)

        if potiontype == len(prefix):
            phrase = "3 AM " + pickword(off) + " potion"
        else:
            phrase = prefix[potiontype] + ' ' + pickword(off) + " potion"
    elif outtype == 5:
        #keep chance of all box types the same
        boxtype = secrets.randbelow(len(prefix) + 2)

        if boxtype == len(prefix) + 1:
            phrase = pickword(off) + " mystery box"
        elif boxtype == len(prefix):
            phrase = pickword(off) + " dybbuk box"
        else:
            phrase = secrets.choice(prefix) + ' ' + pickword(off) + " box"
    elif outtype == 6:
        #keep chance of all challenge types the same
        challengetype = secrets.randbelow(len(prefix) + 4)

        if challengetype == len(prefix) + 3:
            phrase = "3 AM " + pickword(off) + " challenge"
        elif challengetype == len(prefix) + 2:
            phrase = "24 hour " + pickword(off) + " challenge"
        elif challengetype == len(prefix) + 1:
            phrase = "overnight " + pickword(off) + " challenge"
        elif challengetype == len(prefix):
            phrase = "24 hour overnight " + pickword(off) + " challenge"
        else:
            phrase = secrets.choice(prefix) + ' ' + pickword(off) + " challenge"
    else:
        print("ERROR: Invalid phrase type!");
        sys.exit(1)

    return str(phrase)

#open json access token
def readtoken(token):
    try:
        with open(str(token)) as replyfile:
            json_obj = json.load(replyfile)
    except OSError as err:
            print("ERROR:",err,'\n')
            sys.exit(1)
    
    return json_obj["access_token"]

#returns the main version line as a string
#used in version and license commands
def verline():
    return str("Dark Web Mystery Bot v" + progver)

#return part two of version info as string
def verpart2():
    return str("A Fediverse phrase bot by redneonglow.\nMore info: https://github.com/redblade7/darkwebmysterybot")

#shows version info
def optversion():
    print(verline())
    print(verpart2())

#shows license info
def optlicense():
    print(verline())
    print("\nCopyright (c) 2019-22, redneonglow\nAll rights reserved.\n")
    print("Redistribution and use in source and binary forms, with or without\nmodification, are permitted provided that the following conditions are met:\n")
    print("1. Redistributions of source code must retain the above copyright notice, this\n   list of conditions and the following disclaimer.")
    print("2. Redistributions in binary form must reproduce the above copyright notice,\n   this list of conditions and the following disclaimer in the documentation\n   and/or other materials provided with the distribution.\n")
    print("THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \"AS IS\"\nAND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE\nIMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE\nDISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE\nFOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL\nDAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR\nSERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER\nCAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,\nOR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE\nOF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.")

#print phrase to stdout (incl. offensive)
def optprint():
    print(genphrase(True))

#post phrase to fediverse (public)
#vis: Mastodon.py visibility setting (string)
#off: allow offensive content (bool)
def optpostphrase(baseurl,token,vis,off):

    try:
        mastodon = Mastodon(api_base_url=str(baseurl),access_token=readtoken(str(token)))
        mastodon.status_post(genphrase(bool(off)),visibility=str(vis))
    except ValueError as err:
        print("ERROR:",err,'\n')
        sys.exit(2)
    except MastodonError as err:
        print("ERROR:",err,'\n')
        sys.exit(1)

    print("Successfully posted phrase to " + str(baseurl) + '!')

#post version info to fediverse (unlisted)
def optpostver(baseurl,token):

    try:
        mastodon = Mastodon(api_base_url=str(baseurl),access_token=readtoken(str(token)))
        mastodon.status_post(verline()+'\n'+verpart2(),visibility="unlisted")
    except MastodonError as err:
        print("ERROR:",err,'\n')
        sys.exit(1)

    print("Successfully posted version info to " + str(baseurl) + '!')

#main
def main():
   
    parser = argparse.ArgumentParser()
    group = parser.add_mutually_exclusive_group()
    group.add_argument("-l","--license",help="Show license info",action="store_true")
    group.add_argument("-v","--version",action="store_true",help="Show version info")
    group.add_argument("-w","--postversion",help="Post version info once, unlisted, to Fediverse site SERVER using token file TOKEN.",type=str,nargs=2,metavar=("SERVER","TOKEN"))
    group.add_argument("-p","--printphrase",help="Print phrase to stdout NUM times",type=str,metavar="NUM")
    group.add_argument("-o","--postphrase",help="Post phrase once to Fediverse site SERVER using token file TOKEN and visibility value VISIBILITY.",type=str,nargs=3,metavar=("SERVER","TOKEN","VISIBILITY"))
    group.add_argument("-c","--postpcphrase",help="Same as -o, but avoids phrases not suitable for safe-space instances.",type=str,nargs=3,metavar=("SERVER","TOKEN","VISIBILITY"))

    args = parser.parse_args()

    if args.license:
        optlicense()
    elif args.version:
        optversion()
    elif args.postversion:
        optpostver(args.postversion[0],args.postversion[1])
    elif args.postphrase:
        optpostphrase(args.postphrase[0],args.postphrase[1],args.postphrase[2],True)
    elif args.postpcphrase:
        optpostphrase(args.postpcphrase[0],args.postpcphrase[1],args.postpcphrase[2],False)
    elif args.printphrase:

        try:
            times = int(args.printphrase)
        except ValueError: 
            print("ERROR: argument -p/--printphrase NUM must be whole number!\n")
            parser.print_help()
            sys.exit(2)

        if times < 1:
            print("ERROR: argument -p/--printphrase NUM must be 1 or greater!\n")
            parser.print_help()
            sys.exit(2)
        else:
            for count in range(0,times):
                optprint()

    else:
        print("ERROR: Invalid command!\n")
        parser.print_help()
        sys.exit(2)
        
if __name__ == "__main__":
    main()
    sys.exit(0)
